### ISSUE TITLE
cache="ram" mode produces nondeterministic results even with fixed seed

### ISSUE THREAD

**Issue by ShreeSinghi** [ultralytics/ultralytics#21783]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

When training YOLO with cache="ram" and a fixed seed, results are nondeterministic across runs. Although a warning is currently shown, this behavior still breaks makes controlled experiments difficult.

The issue comes from multi-worker buffer loading in the mosaic augmentation pipeline. Since workers finish loading images in a nondeterministic order, the mosaic buffer is initialized differently each run. And this introduces randomness uncontrolled by the seed.

**Proposed fix:**

Reset the buffer once after it has been loaded, before training begins, using an order determined by the given seed. This ensures reproducibility when a seed is set, while retaining the training speed benefits of RAM caching.

### Use case

Researchers often rely on reproducibility for fair benchmarking and ablation studies.
Currently, using cache="ram" makes reproducibility impossible, and fixing this would allow reproducible experiments without sacrificing training time.

### Additional

I can submit a PR with a minimal change to implement this fix.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ShreeSinghi, thanks for reporting this and for offering to contribute a PR 🚀 This is an automated response — an Ultralytics engineer will also follow up with you soon.

For general guidance, we recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a 🐛 Bug Report related to determinism with `cache="ram"` and mosaic buffering, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can verify and debug it. Helpful details:
- Exact command or Python snippet you run (including `cache="ram"`, seed, and any mosaic settings)
- Environment details (OS, Python, PyTorch, CUDA/CUDNN versions), GPU count, and `num_workers`
- Whether you’re using DDP, AMP, or any other training flags
- Full logs from at least two runs showing the nondeterministic behavior
- Dataset sample or small synthetic dataset for the MRE if possible

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Regarding your proposed fix: thank you for offering to submit a PR 🙌 Please include the MRE, a clear description of the change (e.g., how the buffer is reset deterministically before training), and notes on any performance considerations. Adding a small regression check to confirm reproducibility across runs would be very helpful.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the thorough analysis—this aligns with our existing warning that `cache='ram'` can be non-deterministic when `hyp.deterministic=True`; see the BaseDataset logic in the [data/base reference](https://docs.ultralytics.com/reference/data/base/). We’d welcome a PR that, when `cfg.deterministic` is True, deterministically reinitializes the mosaic buffer once (right after `cache_images()` and before training), while preserving current behavior otherwise. Please keep the change minimal, avoid adding new args to `default.yaml`, include a small test demonstrating identical results across two runs with the same `seed` (seeding utils: [init_seeds](https://docs.ultralytics.com/reference/utils/torch_utils/)), and add a brief docs note; contributing guidelines are here: [Contributing to Ultralytics](https://docs.ultralytics.com/help/contributing/).
